### PR TITLE
[Data] Enable sort over multiple keys in datasets

### PR DESCRIPTION
**Pull request by jaidisido**
## Why are these changes needed?

This is a first draft into enabling sorting Ray datasets over multiple keys (i.e. passing a list of keys to `.sort()`).

A new unit test `test_sort_with_multiple_keys` showcases an example.

## Related issue number

https://github.com/ray-project/ray/issues/25732

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [X] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/arrow_block.py
Comment (jaidisido) [inline python/ray/data/_internal/arrow_block.py:435]: `np.searchsorted` does not support multi-dimensional arrays. This implementation over multiple keys is not vectorised and therefore not as optimal/performant. Would appreciate pointers on how to improve it.

File: python/ray/data/_internal/planner/exchange/sort_task_spec.py
Comment (jaidisido) [inline python/ray/data/_internal/planner/exchange/sort_task_spec.py:125]: `sample_boundaries` now returns a list of tuples sorted across the keys. For example:
```
[("A", 1, True), ("A", 2, False), ("B", 1, False)...]
```

**Review by jaidisido — COMMENTED** [review]
Clarifications

**Comment by zhe-thoughts** [comment]
Question from live discussion: is it OK to start with PyArrow blocks and later extend to more generic cases? @raulchen

**Comment by raulchen** [comment]
@zhe-thoughts  I think that is fine as long as the code structure is generic and extensible in the future. E.g., we should define some abstraction in the base class and leave some subclasses as NotImplemented. Will look at this PR later today.

**Comment by kukushking** [comment]
Hi @zhe-thoughts @raulchen fyi this PR also includes pandas blocks implementation now - added on Friday

**Comment by raulchen** [comment]
Sorry for the delay. Just got a chance to fully read through the whole PR today. 

One major concern is that the type definition of the sort key has been been messy and inconsistent across the code base. Of course, this is an existing issue. But extending to multiple sort columns would make the situation worse. I'm wondering if you'd be able to do a sanitization refactor around this (If you don't have time, we can work on this as well).

Here is the current situation:
- For `Dataset.sort`,  `SortStage`, `_validate_key_fn`, and the `Sort` logical op, we use `key: Optinal[str]`, `descending: bool`. 
- But for `generate_sort_fn`, `SortTaskSpec`, `_internal/sort.py`, and `BlockAccessor.sample`, we use `key: SortKeyT`, which is `Union[None, List[Tuple[str, str]], Callable[[T], Any]]`. 
  - For `SortKeyT`,  if it's `List[Tuple[str, str]]`, it'd be something like `[("column1", "ascending"), ("column2", "descending")]`. Because of this, we have a lot of code doing something like `col = key[0][0]` or `isinstance(key, list)`, which is very hard to read. I believe the reason why it's defined as this is to match the definition of `pyarrow.compute.sort_indices`. Also, the Callable case seems to be not implemented at all (@c21 correct me if I'm wrong).

To sanitize this, I'm thinking of the following approach:
- At the API level (only `Dataset.sort`), we still use `key: Optinal[str]`, `descending: bool`. And extend them to lists after this PR.
- For everything else, we should define a `SortKey` class to incapsulate all the related logic. The class will provide methods like `get_columns`, `to_arrow_sort_args`, `to_pandas_sort_args`, etc.

File: python/ray/data/_internal/arrow_block.py
Comment (raulchen) [reply python/ray/data/_internal/arrow_block.py:435]: it's weird that I find we are already using `to_pylist` and our CI also test against pyarrow 6. Not sure why that can work. I'll confirm. also, in this is indeed needed, we can move the definition to `python/ray/data/_internal/arrow_ops/transform_pyarrow.py`.

File: python/ray/data/_internal/arrow_block.py
Comment (raulchen) [inline python/ray/data/_internal/arrow_block.py:436]: This looks heavy. I guess it's possible to bisect on the original arrow table, without converting it to a pylist.

File: python/ray/data/dataset.py
Comment (kukushking) [reply python/ray/data/dataset.py:2018]: Thanks, that makes sense. I was wondering what high-level api that allows multi-directional sort should look like for the user - whether it's a list of tuples like it is in some parts of the existing codebase or something else. `sort_values` is a good example.

Comment (jaidisido) [reply python/ray/data/dataset.py:2018]: Multi-directional sort might be ambitious in my opinion. It would require a completely different approach as `bisect`does not support it

File: python/ray/data/_internal/arrow_block.py
Comment (jaidisido) [reply python/ray/data/_internal/arrow_block.py:436]: Would obviously prefer to bisect on the entire arrow table but not sure how you are suggesting to do it here?

File: python/ray/data/tests/test_sort.py
Comment (jaidisido) [reply python/ray/data/tests/test_sort.py:166]: The test is parameterized for both pandas and arrow blocks, so `map_batches` is used for that block conversion. It was extracted from one of you existing [tests](https://github.com/ray-project/ray/blob/master/python/ray/data/tests/test_sort.py#L84)

File: python/ray/data/_internal/arrow_block.py
Comment (kukushking) [reply python/ray/data/_internal/arrow_block.py:435]: CI for pyarrow 6 was [failing](https://buildkite.com/ray-project/oss-ci-build-pr/builds/28395#018949ee-1589-4c63-8ffe-e4043f11c4e5). Thanks, moved to `arrow_ops`.

**Comment by kukushking** [comment]
Hi @raulchen yeah that makes sense. I'll give the approach a shot.

On another note -  looking at the last ci there's two failures seemingly unrelated to the PR - ray train (`python/ray/train/tests/test_lightning_deepspeed.py::test_deepspeed_stages`) and link check. Do you know what to do with those? In particular train test - is that just flaky, or potentially shows a performance drop?

**Comment by raulchen** [comment]
That seems to be a flakey test, and have been fixed on the latest master. All flakey tests can be found [here](https://flakey-tests.ray.io) (search Lightning for this one).

File: python/ray/data/dataset.py
Comment (raulchen) [reply python/ray/data/dataset.py:2018]: I think that can be done with a custom key function for `bisect`. Did you try that? 
Also by using the custom key function, I think we can avoid reversing the order of the items.

File: python/ray/data/_internal/arrow_block.py
Comment (raulchen) [reply python/ray/data/_internal/arrow_block.py]: I was thinking of doing the bisect with the original arrow table, with a custom key argument.

File: python/ray/data/dataset.py
Comment (jaidisido) [reply python/ray/data/dataset.py:2018]: AFAIK the key argument to bisect was only introduced in Python 3.10+, so isn't available in 3.8, 3.9
https://docs.python.org/3.10/library/bisect.html
https://docs.python.org/3.8/library/bisect.html

**Comment by kukushking** [comment]
Thanks @raulchen added recommended sort key implementation in [commit](https://github.com/kukushking/ray/commit/3546444cb205abd94462d6b304f242d974629812).

**Comment by raulchen** [comment]
> Thanks @raulchen added recommended sort key implementation in [commit](https://github.com/kukushking/ray/commit/3546444cb205abd94462d6b304f242d974629812).

Nice. Could you submit a PR for this change?

File: python/ray/data/tests/test_sort.py
Comment (raulchen) [reply python/ray/data/tests/test_sort.py:166]: I see.

File: python/ray/data/dataset.py
Comment (raulchen) [reply python/ray/data/dataset.py:2018]: Got it. But then I think it's worth it to implement our own bisect to avoid copy.

**Comment by kukushking** [comment]
Thanks @raulchen. Before I merge, after having gone through this refactor (latest working version that passes all sort tests is [here](https://github.com/kukushking/ray/commit/bf407bba6e0d6909fa42362bbf2bf7c090099214)) there's a few details I'd like to discuss:
- It looks like what masks as sort key inconsistency in some cases is caused by a difference between user-facing and internal APIs. By user-facing I mean `Dataset.sort` but also I think we should include `BlockAccessor.sort_and_partition`, and `merge_sorted_blocks`, `sample`, including `BlockAccessor` subclasses. As a consequence,`SortStage`, and `SortOp` should also use the same definition, although we can use `SortKeyT` type var as a shortcut here - I think this was the intent. 
- Internally, however, block implementations (`arrow_block`, `pandas_block`, `table_block`) should use the sort key class you proposed to convert between the pandas/arrow/etc sort args.

Let me know what you think.

**Comment by raulchen** [comment]
hey @kukushking, the code looks good to me. But I don't fully understand your proposal. Are you proposing to keep `SortKeyT` for some APIs?

**Comment by kukushking** [comment]
@raulchen Yes that's right. It seems that `BlockAccessor` methods are also user-facing so it would make sense to keep the same definition as `Dataset.sort` so wanted to hear your opinion here. Happy to merge my branch into this PR as-is though.

File: python/ray/data/dataset.py
Comment (kukushking) [reply python/ray/data/dataset.py:2018]: Custom multi-order key function would work well if it's just simple types like ints i.e. `lambda x: x[0], -x[1]` but it gets more complicated with other types.

**Comment by raulchen** [comment]
@kukushking I see. But `BlockAccessor` isn't supposed to be user-facing. If any docs/comments cause the confusion, please let us know. We should update them.  By the way, if possible, it'd be more preferable to submit the refactor change as a separate PR, so the change is easier to review and track. But if that will bring significant overheads, using 1 PR is acceptable as well.

**Comment by kukushking** [comment]
@raulchen sounds good, I will submit separate PR. Thanks!

**Comment by PhysicsACE** [comment]
@raulchen I was going to submit this PR before but due to hardware issues, I was unable to access my local repo as I did not have access to my computer. I have a working prototype of multikey sorting/grouping/aggregating on my forked branch https://github.com/PhysicsACE/ray/tree/multikey and if it looks good, I would be happy to merge the changes into this PR for the requested feature.

**Comment by raulchen** [comment]
> @raulchen I was going to submit this PR before but due to hardware issues, I was unable to access my local repo as I did not have access to my computer. I have a working prototype of multikey sorting/grouping/aggregating on my forked branch https://github.com/PhysicsACE/ray/tree/multikey and if it looks good, I would be happy to merge the changes into this PR for the requested feature.

Sorry, I am confused. How is that branch different from this PR? Do you mean that branch not only implements sort, but also groupby and agg? I think it'd be better to submit different PRs if possible.

**Comment by PhysicsACE** [comment]
> > @raulchen I was going to submit this PR before but due to hardware issues, I was unable to access my local repo as I did not have access to my computer. I have a working prototype of multikey sorting/grouping/aggregating on my forked branch https://github.com/PhysicsACE/ray/tree/multikey and if it looks good, I would be happy to merge the changes into this PR for the requested feature.
> 
> Sorry, I am confused. How is that branch different from this PR? Do you mean that branch not only implements sort, but also groupby and agg? I think it'd be better to submit different PRs if possible.

It implements both sort and groupby. It is different than this branch in terms of implementation as it has a custom searchsorted function to support directions for each column that can be integrated with this PR. Once sort is supported, I can create a separate PR for groupby as it requires the same sort_and_partition function.

**Comment by raulchen** [comment]
> > > @raulchen I was going to submit this PR before but due to hardware issues, I was unable to access my local repo as I did not have access to my computer. I have a working prototype of multikey sorting/grouping/aggregating on my forked branch https://github.com/PhysicsACE/ray/tree/multikey and if it looks good, I would be happy to merge the changes into this PR for the requested feature.
> > 
> > 
> > Sorry, I am confused. How is that branch different from this PR? Do you mean that branch not only implements sort, but also groupby and agg? I think it'd be better to submit different PRs if possible.
> 
> It implements both sort and groupby. It is different than this branch in terms of implementation as it has a custom searchsorted function to support directions for each column that can be integrated with this PR. Once sort is supported, I can create a separate PR for groupby as it requires the same sort_and_partition function.

I see. thanks for the clarification. Maybe let's wait until this PR is merged, and then you can submit a separate one?

File: python/ray/data/dataset.py
Comment (raulchen) [inline python/ray/data/dataset.py:2178]: typo columns

File: python/ray/data/_internal/sort.py
Comment (raulchen) [inline python/ray/data/_internal/sort.py:62]: this error message is a bit misleading. just say lengths don't match?

File: python/ray/data/_internal/planner/exchange/sort_task_spec.py
Comment (raulchen) [reply python/ray/data/_internal/planner/exchange/sort_task_spec.py:115]: Could you add this to the function comment? and could you also add some comments explaining the above code? It's not very easy to understand.

File: python/ray/data/dataset.py
Comment (kukushking) [reply python/ray/data/dataset.py]: Thank you!

File: python/ray/data/_internal/planner/exchange/sort_task_spec.py
Comment (kukushking) [reply python/ray/data/_internal/planner/exchange/sort_task_spec.py:121]: Thanks, that dict/list comprehension was a bit heavy. I've refactored it & removed sorting inside of the loop. Added clarifying comments as well.

**Review by raulchen — APPROVED** [review]

**Pull request merged**
